### PR TITLE
Implement Sync Rules

### LINKED ISSUE #15: Implement sync rules
Once a Metadata object sync has been defined between the master and slave it should be stored in the data store and the application should be able to pull up this record through a search so that it is not required to redefine the sync configuration for future edits. There could be an option to sync once off or store the sync rule.

### PR DESCRIPTION
**Pull request by SferaDev**
### :pushpin: References

* **Issue:** Close #15 

### :memo: Implementation

- Add two new routes to access the new pages

- Add Sync Rules Landing Page

- Add Sync Rules Creation Wizard

    - Add General Info Step with a simple form

    - Add Metadata Selection Step with ObjectsTable that allows any type of metadata

    - Add Instance Selection Step with a multiselector

    - Add Save Step with a Summary of the Sync Rule to save in the dataStore

- Update dropdown to allow hiding default value and fixing bug where it goes off-screen

- Move getValueForCollection method to utils and re-use it in the new components

- Fix bug in datastore that crashed when searching in non-strings

- Add a new list by id logic method that does not require knowing the metadata type

- Allow setting the fields when accessing the metadata endpoint

### :art: Screenshots

image

image

image

image

image

image

image

image

image

### :bookmark_tabs: Others

- Any dependencies from other projects?
    - https://github.com/EyeSeeTea/d2-ui-components/pull/41
    - https://github.com/EyeSeeTea/d2-ui-components/pull/42
    - https://github.com/EyeSeeTea/d2-ui-components/pull/43
    - https://github.com/EyeSeeTea/d2-ui-components/pull/44

### CODE REVIEW

File: src/components/rules-creation-page/steps/MetadataSelectionStep.jsx [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+                autoHideDuration: 1000,
+            });
+        } else if (difference < 0) {
+            snackbar.info(
+                i18n.t("Removed {{difference}} elements", { difference: Math.abs(difference) }),
+                { autoHideDuration: 1000 }
+            );
+        }
+
+        syncRule.selectedIds = selectedIds;
```
Comment (tokland) [inline src/components/rules-creation-page/steps/MetadataSelectionStep.jsx]: Same

File: src/models/syncRule.ts [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+            name: "",
+            description: "",
+            originInstance: "",
+            builder: {
+                targetInstances: [],
+                metadata: {},
+            },
+            selectedIds: []
+        });
+    }
```
[the remark this answers is not in the document]
Comment (SferaDev) [reply src/models/syncRule.ts]: We could remove selectedIds and calculate them from Metadata. And no, names do not need to be persisted. Will update accordingly this afternoon

File: src/components/rules-creation-page/steps/MetadataSelectionStep.jsx [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+        },
+        {
+            name: this.props.d2.models["indicator"].displayName,
+            id: this.props.d2.models["indicator"].name,
+        },
+        {
+            name: this.props.d2.models["dataElement"].displayName,
+            id: this.props.d2.models["dataElement"].name,
+        },
+    ];
```
[the remark this answers is not in the document]
Comment (tokland) [reply src/components/rules-creation-page/steps/MetadataSelectionStep.jsx]: I guess the idea was to keep the same structure in the app and the persisted one. So let's do this only if it really makes sense to reduce the size of the persisted object.

[the remark this answers is not in the document]
Comment (SferaDev) [reply src/components/rules-creation-page/steps/MetadataSelectionStep.jsx]: Will do in another PR, we might extract the models altogether and pass them as a prop

File: src/components/rules-creation-page/steps/InstanceSelectionStep.jsx [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+});
+
+const InstanceSelectionStep = props => {
+    const { d2, syncRule } = props;
+    const [instanceOptions, setInstanceOptions] = useState([]);
+    const [selectedOptions, setSelectedOptions] = useState(syncRule.targetInstances);
+
+    const changeInstances = instances => {
+        setSelectedOptions(instances);
+        syncRule.targetInstances = instances;
```
[the remark this answers is not in the document]
Comment (SferaDev) [reply src/components/rules-creation-page/steps/InstanceSelectionStep.jsx]: We already receive syncRule as a prop, not sure that approach.

```syncRule.targetInstances``` is actually a setter method of TS but I get your point.

[the remark this answers is not in the document]
Comment (tokland) [reply src/components/rules-creation-page/steps/InstanceSelectionStep.jsx]: As a rule of thumb, a function should not alter its arguments. The arguments of a React component are its props. 

I see that `SyncRule` has imperative setters, while `Instance` has immutable setters. With immutable setters, we can do the typical: `props.onChange(syncRule.setTargetInstances(instances))`.

No need to change it if we anticipate no problems, but in general, it's better to keep props immutable.

[the remark this answers is not in the document]
Comment (SferaDev) [reply src/components/rules-creation-page/steps/InstanceSelectionStep.jsx]: Got it, implemented in an immutable way!

File: src/models/syncRule.ts [hunk 5]
```diff
@@ -59,8 +75,14 @@ export default class SyncRule {
...
-    public updateTargetInstances(instances: string[]): void {
-        this.syncRule.builder.targetInstances = instances;
+    public updateTargetInstances(targetInstances: string[]): SyncRule {
+        return SyncRule.build({
+            ...this.syncRule,
+            builder: {
+                ...this.syncRule.builder,
+                targetInstances,
+            },
+        });
```
Comment (tokland) [inline src/models/syncRule.ts]: Cool! I know at the beginning it feels cumbersome, but in my experience it pays off, on the long run, to have fully immutable models. By the way, this kind of boilerplate for nested updates can be simplified using lodash.fp, _[.set](https://gist.github.com/jfmengels/6b973b69c491375117dc#_setpath-value-object), but please don't change it, the method is not fully typed and we lose safety. In the function world, to work with nested structures, there is a concept called "lenses", for TS this is the best I know: https://github.com/utatti/lens.ts. Again, don't change it, just as a note, it's perfectly fine as it is, with just a level of nesting.

Comment (SferaDev) [reply src/models/syncRule.ts]: WIll take a look on this lenses concept!

**Comment by adrianq** [comment]
First of all, let me say that this incredible work! Thanks for this. A few minor comments (some of them already discussed):
- List of target instances should display as separated commas list
- Left click on a row forward to the edit feature. While in the other sync section it goes to Details. Although in latest versions of DHIS2 the first option is the implemented one, in general, unless you have a strong opinion in a different sense I will choose 'Details' as the left click action.
- Remove asterisk and use compulsory fields instead
- If you click on 'Clear Selection' probably is better to clear up all the filters (including the selected checkbox) and refresh, right?

I am merging this anyway as it is a huge PR and I don't want to postpone it anymore. Feel free to create issues if you agree with my comments or let me know otherwise!

**Pull request merged**